pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,200 @@
+	"go.uber.org/atomic"
+)
+
+type observation struct {
+	lvalues    []string
+	lastUpdate *atomic.Int64
+}
+
+func (t *Tracker) cleanupTrackerAttribution(vals []string) {
+	t.activeSeriesPerUserAttribution.DeleteLabelValues(vals...)
+	t.receivedSamplesAttribution.DeleteLabelValues(vals...)
+	t.discardedSampleAttribution.DeleteLabelValues(vals...)
+}
+
+func (t *Tracker) cleanupTracker(userID string) {
+	filter := prometheus.Labels{"user": userID}
+	t.activeSeriesPerUserAttribution.DeletePartialMatch(filter)
+	t.receivedSamplesAttribution.DeletePartialMatch(filter)
+	t.discardedSampleAttribution.DeletePartialMatch(filter)
+}
+
+type Tracker struct {
+	userID                         string
+	trackedLabels                  []string
+	maxCardinality                 int
+	activeSeriesPerUserAttribution *prometheus.GaugeVec
+	receivedSamplesAttribution     *prometheus.CounterVec
+	discardedSampleAttribution     *prometheus.CounterVec
+
+	// oLock protects the observed map
+	oLock    sync.RWMutex
+	observed map[uint64]*observation
+
+	hashBuffer []byte
+}
+
+func (t *Tracker) IncrementActiveSeries(lbs labels.Labels, now time.Time) {
+	vals := t.getKeyValues(lbs, now.Unix())
+	t.activeSeriesPerUserAttribution.WithLabelValues(vals...).Inc()
+}
+
+func (t *Tracker) IncrementDiscardedSamples(lbs labels.Labels, value float64, reason string, now time.Time) {
+	vals := t.getKeyValues(lbs, now.Unix())
+	t.discardedSampleAttribution.WithLabelValues(vals...).Add(value)
+}
+
+func (t *Tracker) IncrementReceivedSamples(lbs labels.Labels, value float64, now time.Time) {
+	vals := t.getKeyValues(lbs, now.Unix())
+	t.receivedSamplesAttribution.WithLabelValues(vals...).Add(value)
+}
+
+func (t *Tracker) getKeyValues(lbls labels.Labels, ts int64) []string {
+	values := make([]string, len(t.trackedLabels)+1)
+	for i, l := range t.trackedLabels {
+		values[i] = lbls.Get(l)
+		if values[i] == "" {
+			values[i] = missingValue
+		}
+	}
+	values[len(values)-1] = t.userID
+
+	var stream uint64
+	stream, t.hashBuffer = lbls.HashForLabels(t.hashBuffer, t.trackedLabels...)
+	if t.overflow(stream, values, ts) {
+		// Omit last label.
+		for i := range values[:len(values)-1] {
+			values[i] = overflowValue
+		}
+	}
+
+	return values
+}
+
+func (t *Tracker) overflow(stream uint64, values []string, ts int64) bool {
+	// If the maximum cardinality is hit all streams become `__overflow__`.
+	if len(t.observed) > t.maxCardinality {
+		return true
+	}
+
+	if o, known := t.observed[stream]; known && o.lastUpdate != nil && o.lastUpdate.Load() < ts {
+		o.lastUpdate.Store(ts)
+	} else {
+		t.observed[stream] = &observation{
+			lvalues:    values,
+			lastUpdate: atomic.NewInt64(ts),
+		}
+	}
+
+	return false
+}
+
+// we need the time stamp, since active series could have entered active stripe long time ago, and already evicted
+// from the observed map but still in the active Stripe
+func (t *Tracker) DecrementActiveSeries(lbs labels.Labels, value int64, ts time.Time) {
```

Comment:
I think here we might still want ts, when we decrement a active series counter that are "active" we may want to keep the timestamp updated.

Reply:
Why would we want to update the timestamp of something that didn't have activity?